### PR TITLE
userguide: explain rule types and categorization - v4

### PR DESCRIPTION
**Pull request by jufajardini**
Add documentation about the rule types introduced by 2696fda04168cb82.

Add doc tags around code definitions that are referenced in the docs.

Link to ticket: https://redmine.openinfosecfoundation.org/issues/
I guess this covers https://redmine.openinfosecfoundation.org/issues/7031

Built docs: https://suri-rtd-test.readthedocs.io/en/doc-sigtypes-et-properties-v4/rules/intro.html#rule-s-types-and-categorization

Previous PR: https://github.com/OISF/suricata/pull/12107

Describe changes: (tried to) address feedback:
- better explain `Scope`, and tie explanations to other Table columns as well
- add code excerpts for `SignatureType` and `SignatureProperties`
- Fix explanations that were wrong or unclear
- Get a stab at trying to explain when do `app-layer protocol` rules are inspected
- associate commit with ticket that seems related

### CODE REVIEW

File: doc/userguide/rules/intro.rst [hunk 1]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
+        (packet payload and stream payload)
+      - 'content' with 'starts with' or 'depth'
+    * - Stream
+      - Flow, if stateful (**)
+      - Per stream chunk, if stateful, per-packet if not
+      - Against the reassembled stream. If stream unavailable, match per-packet
+      - 'tcp-stream' in protocol field; simple 'content'; 'byte_extract'
+    * - Application Layer Protocol
+      - Flow
+      - Per-packet basis
```
Comment (jufajardini) [inline doc/userguide/rules/intro.rst:434]: Is this more correct?

File: doc/userguide/rules/intro.rst [hunk 2]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
+    * - IP Only
+      - Flow
+      - Once per direction
+      - On IP addresses on the flow
+      - Source/ Destination field of a rule
+    * - IP Only (contains a negated address)(*)
+      - Flow
+      - Once per direction
+      - On the flow, on IP address level (negated addresses)
+      - Source/ Destination field of a rule, containing negated address
```
Comment (jufajardini) [inline doc/userguide/rules/intro.rst:415]: Added this, not sure if good to have or not

File: doc/userguide/rules/intro.rst [hunk 3]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
 See :ref:`rules-http-uri-normalization` for more information.
+
+
+Rule Types and Categorization
+-----------------------------
+
+Once parsed, Suricata rules are categorized for performance and further
+processing (as different rule types will be handled by specific engine modules).
+The signature types are defined in `src/detect.h
+<https://github.com/OISF/suricata/blob/master/src/detect.h>`_:
```
Comment (jufajardini) [inline doc/userguide/rules/intro.rst:360]: removed the line references here, to make this more time resilient

File: doc/userguide/rules/intro.rst [hunk 4]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
+.. container:: example-rule
+
+    alert tcp-stream any any -> any any (msg:"tcp-stream, no content"; sid:101;)
+
+
+.. container:: example-rule
+
+    alert tcp-pkt [192.168.0.0/16,10.0.0.0/8,172.16.0.0/12] any -> any any (msg:"tcp-pkt, no content"; sid:201;)
+
+IP Only (negated address)
```
Comment (jufajardini) [inline doc/userguide/rules/intro.rst:503]: Should it also be `contains negated address` here, to ensure people are seeing it?

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23303

File: doc/userguide/rules/intro.rst [hunk 5]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
+   :alt: A flowchart representing the SignatureSetType function.
+
+The following table lists all Suricata signature types, and how they impact the
+aspects aforementioned.
+
+.. list-table:: Suricata Rule Types
+    :header-rows: 1
+
+    * - Type
+      - Scope
```
Comment (victorjulien) [inline doc/userguide/rules/intro.rst:392]: action scope?

File: doc/userguide/rules/intro.rst [hunk 6]
```diff
@@ -349,3 +349,220 @@ reassembled streams, TLS-, SSL-, SSH-, FTP- and dcerpc-buffers.
...
+      - Flow
+      - Once per direction, when protocol detection is done
+      - On protocol detected for the flow
+      - app-layer-protocol
+    * - Packet-Stream
+      - Flow, if stateful (**)
+      - Flow, if stateful, per-packet if not
+      - Against the reassembled stream. If stream unavailable, match per-packet
+        (packet payload and stream payload)
+      - 'content' with 'starts with' or 'depth'
```
Comment (victorjulien) [inline doc/userguide/rules/intro.rst:426]: startswith is a keyword w/o space

File: doc/userguide/rules/intro.rst (hunk 1 above)
Comment (victorjulien) [reply doc/userguide/rules/intro.rst:434]: Think so, ya

File: doc/userguide/rules/intro.rst (hunk 4 above)
Comment (victorjulien) [reply doc/userguide/rules/intro.rst:503]: perhaps another example as well, e.g. `[10.0.0.0/8,!10.10.10.10]`

**Comment by jufajardini** [comment]
New PR incoming to fix clang failure

**Review by victorjulien — CHANGES_REQUESTED** [review]
minor comments inline

File: doc/userguide/rules/intro.rst (hunk 6 above)
Comment (jufajardini) [reply doc/userguide/rules/intro.rst:426]: >_<'

**Comment by jufajardini** [comment]
New version: https://github.com/OISF/suricata/pull/12112

**Pull request closed without merging**